File: src/server/gateway_server.py
from __future__ import annotations
import asyncio
import time
import uuid
from dataclasses import dataclass
from typing import Any, Optional

import uvicorn
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from fastapi.responses import HTMLResponse, JSONResponse
from pydantic import ValidationError

from ..protocol.frames import (
    ErrorShape,
    EventFrame,
    FeaturesInfo,
    HelloOk,
    RequestFrame,
    ResponseFrame,
    ServerInfo,
)
from ..sessions.session_manager import SessionManager
from ..agent.agent_runtime import AgentRuntime, AgentTurnEvent

PROTOCOL_VERSION = 1
SERVER_VERSION = "0.1.0"

METHODS = [
    "chat.send",
    "sessions.list",
    "sessions.create",
    "sessions.delete",
    "sessions.resolve",
    "sessions.history",
]
EVENTS = ["chat.delta", "chat.message", "chat.idle", "chat.tool", "tick"]


@dataclass
class _Client:
    conn_id: str
    ws: WebSocket
    authenticated: bool = False
    subscribed_session: Optional[str] = None


class GatewayServer:
    def __init__(
        self,
        port: int,
        session_manager: SessionManager,
        agent_runtime: AgentRuntime,
        token: Optional[str] = None,
        heartbeat_interval_ms: int = 30000,
    ) -> None:
        self._port = port
        self._token = token
        self._heartbeat_ms = heartbeat_interval_ms
        self._sessions = session_manager
        self._agent = agent_runtime
        self._clients: dict[str, _Client] = {}
        self._event_seq = 0
        self._heartbeat_task: Optional[asyncio.Task] = None
        self._server: Optional[uvicorn.Server] = None

        self.app = FastAPI()
        self.app.add_api_route("/health", self._health, methods=["GET"])
        self.app.add_api_route("/", self._info, methods=["GET"])
        self.app.add_api_websocket_route("/", self._handle_connection)

    # ── HTTP handlers ─────────────────────────────────────────────────────────

    async def _health(self) -> JSONResponse:
        return JSONResponse({"status": "ok", "ts": time.time()})

    async def _info(self) -> HTMLResponse:
        return HTMLResponse(
            f"<html><body><h1>Revolut Gateway</h1>"
            f"<p>WebSocket: <code>ws://localhost:{self._port}/</code></p>"
            f"<p>Health: <a href='/health'>/health</a></p></body></html>"
        )

    # ── Lifecycle ─────────────────────────────────────────────────────────────

    async def start(self) -> None:
        self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())
        config = uvicorn.Config(
            self.app,
            host="0.0.0.0",
            port=self._port,
            log_level="warning",
            access_log=False,
        )
        self._server = uvicorn.Server(config)
        await self._server.serve()

    async def stop(self) -> None:
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
        for client in list(self._clients.values()):
            try:
                await client.ws.close()
            except Exception:
                pass
        if self._server:
            self._server.should_exit = True

    # ── Heartbeat ─────────────────────────────────────────────────────────────

    async def _heartbeat_loop(self) -> None:
        while True:
            await asyncio.sleep(self._heartbeat_ms / 1000)
            await self._broadcast_event("tick", {"ts": time.time()})

    # ── Connection handling ───────────────────────────────────────────────────

    async def _handle_connection(self, ws: WebSocket) -> None:
        await ws.accept()
        conn_id = uuid.uuid4().hex[:12]
        client = _Client(conn_id=conn_id, ws=ws)

        # First message must be a connect request
        try:
            raw = await asyncio.wait_for(ws.receive_text(), timeout=30)
        except (asyncio.TimeoutError, WebSocketDisconnect):
            await ws.close()
            return

        try:
            frame = RequestFrame.model_validate_json(raw)
        except (ValidationError, Exception):
            await self._send_error(ws, None, "PARSE_ERROR", "Failed to parse frame")
            await ws.close()
            return

        if frame.method != "connect":
            await self._send_error(ws, frame.id, "PROTOCOL_ERROR", "First message must be connect")
            await ws.close()
            return

        if self._token:
            params = frame.params or {}
            auth_token = params.get("auth", {}).get("token", "") if isinstance(params, dict) else ""
            if auth_token != self._token:
                await self._send_error(ws, frame.id, "AUTH_ERROR", "Invalid or missing token")
                await ws.close()
                return

        client.authenticated = True
        self._clients[conn_id] = client

        hello = HelloOk(
            protocol=PROTOCOL_VERSION,
            server=ServerInfo(version=SERVER_VERSION, connId=conn_id),
            features=FeaturesInfo(methods=METHODS, events=EVENTS),
        )
        await self._send_response(ws, frame.id, True, hello.model_dump())

        try:
            async for raw_msg in ws.iter_text():
                await self._handle_message(client, raw_msg)
        except WebSocketDisconnect:
            pass
        finally:
            self._clients.pop(conn_id, None)

    async def _handle_message(self, client: _Client, raw: str) -> None:
        try:
            frame = RequestFrame.model_validate_json(raw)
        except (ValidationError, Exception):
            await self._send_error(client.ws, None, "PARSE_ERROR", "Failed to parse frame")
            return
        try:
            payload = await self._route_method(client, frame)
            await self._send_response(client.ws, frame.id, True, payload)
        except Exception as e:
            await self._send_error(client.ws, frame.id, "METHOD_ERROR", str(e))

    # ── Method routing ────────────────────────────────────────────────────────

    async def _route_method(self, client: _Client, frame: RequestFrame) -> Any:
        params: dict = frame.params if isinstance(frame.params, dict) else {}
        match frame.method:
            case "chat.send":
                return await self._handle_chat_send(client, params)
            case "sessions.list":
                return {"sessions": [s.__dict__ for s in self._sessions.list_sessions()]}
            case "sessions.create":
                entry = await self._sessions.create_session(
                    params.get("key", "default"), params.get("title")
                )
                return entry.__dict__
            case "sessions.delete":
                deleted = await self._sessions.delete_session(params.get("key", ""))
                return {"deleted": deleted}
            case "sessions.resolve":
                entry = await self._sessions.resolve_or_create(
                    params.get("key", "default"), params.get("title")
                )
                return entry.__dict__
            case "sessions.history":
                messages = await self._sessions.read_transcript(params.get("key", ""))
                return {"messages": [m.__dict__ for m in messages]}
            case _:
                raise ValueError(f"Unknown method: {frame.method}")

    async def _handle_chat_send(self, client: _Client, params: dict) -> dict:
        session_key = params.get("key", "default")
        message = params.get("message", "")
        await self._sessions.resolve_or_create(session_key)
        client.subscribed_session = session_key

        async def run_agent() -> None:
            def on_event(event: AgentTurnEvent) -> None:
                asyncio.create_task(self._dispatch_agent_event(session_key, event))

            try:
                await self._agent.run_turn(session_key, message, on_event)
            except Exception as e:
                await self._dispatch_agent_event(
                    session_key, AgentTurnEvent(type="error", error=str(e))
                )

        asyncio.create_task(run_agent())
        return {"status": "queued"}

    async def _dispatch_agent_event(self, session_key: str, event: AgentTurnEvent) -> None:
        match event.type:
            case "delta":
                await self._broadcast_session_event(
                    session_key, "chat.delta", {"delta": event.content, "key": session_key}
                )
            case "message":
                await self._broadcast_session_event(
                    session_key, "chat.message", {"content": event.content, "key": session_key}
                )
            case "idle":
                await self._broadcast_session_event(
                    session_key, "chat.idle", {"key": session_key}
                )
            case "error":
                await self._broadcast_session_event(
                    session_key, "chat.error", {"error": event.error, "key": session_key}
                )

    # ── Send helpers ──────────────────────────────────────────────────────────

    async def _send_response(
        self, ws: WebSocket, req_id: Optional[str], ok: bool, payload: Any = None
    ) -> None:
        frame = ResponseFrame(id=req_id or "", ok=ok, payload=payload)
        await ws.send_text(frame.model_dump_json())

    async def _send_error(
        self, ws: WebSocket, req_id: Optional[str], code: str, message: str
    ) -> None:
        frame = ResponseFrame(
            id=req_id or "",
            ok=False,
            error=ErrorShape(code=code, message=message),
        )
        await ws.send_text(frame.model_dump_json())

    def _next_seq(self) -> int:
        self._event_seq += 1
        return self._event_seq

    async def _broadcast_event(self, event: str, payload: Any) -> None:
        frame = EventFrame(event=event, payload=payload, seq=self._next_seq())
        data = frame.model_dump_json()
        for client in list(self._clients.values()):
            if client.authenticated:
                try:
                    await client.ws.send_text(data)
                except Exception:
                    pass

    async def _broadcast_session_event(
        self, session_key: str, event: str, payload: Any
    ) -> None:
        frame = EventFrame(event=event, payload=payload, seq=self._next_seq())
        data = frame.model_dump_json()
        for client in list(self._clients.values()):
            if client.authenticated and client.subscribed_session == session_key:
                try:
                    await client.ws.send_text(data)
                except Exception:
                    pass

File: src/server/gateway-server.ts
import { createServer, type IncomingMessage, type Server as HttpServer } from "node:http";
import crypto from "node:crypto";
import { WebSocketServer, WebSocket } from "ws";
import type { SessionManager } from "../sessions/session-manager.js";
import type { AgentRuntime, AgentTurnCallback } from "../agent/agent-runtime.js";
import {
  validateConnectParams,
  validateRequestFrame,
  formatValidationErrors,
  type ConnectParams,
  type RequestFrame,
  type ResponseFrame,
  type EventFrame,
  type HelloOk,
} from "../protocol/index.js";

const PROTOCOL_VERSION = 1;
const SERVER_VERSION = "0.1.0";

const METHODS = [
  "chat.send",
  "sessions.list",
  "sessions.create",
  "sessions.delete",
  "sessions.resolve",
  "sessions.history",
] as const;

const EVENTS = ["chat.delta", "chat.message", "chat.idle", "chat.tool", "tick"] as const;

export interface GatewayServerOptions {
  port: number;
  token?: string;
  heartbeatIntervalMs: number;
  sessionManager: SessionManager;
  agentRuntime: AgentRuntime;
}

interface GatewayClient {
  connId: string;
  ws: WebSocket;
  authenticated: boolean;
  subscribedSession?: string;
}

export class GatewayServer {
  private httpServer: HttpServer;
  private wss: WebSocketServer;
  private clients = new Map<string, GatewayClient>();
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private eventSeq = 0;
  private opts: GatewayServerOptions;

  constructor(opts: GatewayServerOptions) {
    this.opts = opts;

    this.httpServer = createServer((req, res) => {
      if (req.url === "/health") {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ ok: true, clients: this.clients.size }));
        return;
      }
      // Serve a simple redirect to the UI for the root path
      if (req.url === "/" || req.url === "") {
        res.writeHead(200, { "Content-Type": "text/html" });
        res.end(`<html><body><h1>Revolut Gateway</h1><p>WebSocket endpoint: ws://localhost:${opts.port}</p></body></html>`);
        return;
      }
      res.writeHead(404);
      res.end("Not Found");
    });

    this.wss = new WebSocketServer({ server: this.httpServer });
    this.wss.on("connection", (ws, req) => this.handleConnection(ws, req));
  }

  async start(): Promise<void> {
    return new Promise((resolve) => {
      this.httpServer.listen(this.opts.port, () => {
        console.log(`[gateway] Listening on port ${this.opts.port}`);
        this.startHeartbeat();
        resolve();
      });
    });
  }

  async stop(): Promise<void> {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
    for (const client of this.clients.values()) {
      client.ws.close(1001, "Gateway shutting down");
    }
    this.clients.clear();

    return new Promise((resolve) => {
      this.wss.close(() => {
        this.httpServer.close(() => resolve());
      });
    });
  }

  private startHeartbeat(): void {
    this.heartbeatTimer = setInterval(() => {
      this.broadcastEvent("tick", { ts: Date.now() });
    }, this.opts.heartbeatIntervalMs);
  }

  private handleConnection(ws: WebSocket, req: IncomingMessage): void {
    const connId = crypto.randomUUID().slice(0, 12);
    const client: GatewayClient = { connId, ws, authenticated: false };

    // First message must be a connect handshake
    ws.once("message", (data) => {
      console.log(`[gateway] Raw Handshake from ${connId}:`, data.toString());
      try {
        const raw = JSON.parse(data.toString());
        if (!raw || raw.type !== "req" || raw.method !== "connect") {
          this.sendError(ws, raw?.id ?? "0", "PROTOCOL_ERROR", "First message must be a connect request");
          ws.close(4001, "Expected connect");
          return;
        }

        const params = raw.params as ConnectParams;
        if (!validateConnectParams(params)) {
          this.sendError(ws, raw.id, "VALIDATION_ERROR", formatValidationErrors(validateConnectParams.errors));
          ws.close(4002, "Invalid connect params");
          return;
        }

        // Token auth check
        if (this.opts.token && params.auth?.token !== this.opts.token) {
          this.sendError(ws, raw.id, "AUTH_ERROR", "Invalid or missing token");
          ws.close(4003, "Unauthorized");
          return;
        }

        client.authenticated = true;
        this.clients.set(connId, client);

        // Send hello-ok
        const hello: HelloOk = {
          type: "hello-ok",
          protocol: PROTOCOL_VERSION,
          server: { version: SERVER_VERSION, connId },
          features: {
            methods: [...METHODS],
            events: [...EVENTS],
          },
        };

        const res: ResponseFrame = {
          type: "res",
          id: raw.id,
          ok: true,
          payload: hello,
        };
        const resStr = JSON.stringify(res);
        console.log(`[gateway] -> Sending to ${connId}:`, resStr);
        ws.send(resStr);
        console.log(`[gateway] Client connected: ${connId} (${params.client.id})`);

        // Now listen for further messages
        ws.on("message", (msg) => {
          console.log(`[gateway] <- Received from ${connId}:`, msg.toString());
          this.handleMessage(client, msg.toString());
        });
      } catch (err) {
        this.sendError(ws, "0", "PARSE_ERROR", "Invalid JSON");
        ws.close(4000, "Parse error");
      }
    });

    ws.on("close", () => {
      this.clients.delete(connId);
      console.log(`[gateway] Client disconnected: ${connId}`);
    });

    ws.on("error", (err) => {
      console.error(`[gateway] WebSocket error for ${connId}:`, err.message);
    });
  }

  private async handleMessage(client: GatewayClient, raw: string): Promise<void> {
    let frame: RequestFrame;
    try {
      const parsed = JSON.parse(raw);
      if (parsed.type !== "req") return; // ignore non-request frames from client
      if (!validateRequestFrame(parsed)) {
        this.sendError(client.ws, parsed?.id ?? "0", "VALIDATION_ERROR", "Invalid request frame");
        return;
      }
      frame = parsed as RequestFrame;
    } catch {
      this.sendError(client.ws, "0", "PARSE_ERROR", "Invalid JSON");
      return;
    }

    try {
      const result = await this.routeMethod(client, frame);
      const res: ResponseFrame = {
        type: "res",
        id: frame.id,
        ok: true,
        payload: result,
      };
      const resStr = JSON.stringify(res);
      console.log(`[gateway] -> Sending to ${client.connId}:`, resStr);
      client.ws.send(resStr);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      this.sendError(client.ws, frame.id, "METHOD_ERROR", message);
    }
  }

  private async routeMethod(client: GatewayClient, frame: RequestFrame): Promise<unknown> {
    const params = frame.params as Record<string, unknown> | undefined;

    switch (frame.method) {
      case "chat.send":
        return this.handleChatSend(client, params);
      case "sessions.list":
        return this.handleSessionsList();
      case "sessions.create":
        return this.handleSessionsCreate(params);
      case "sessions.delete":
        return this.handleSessionsDelete(params);
      case "sessions.resolve":
        return this.handleSessionsResolve(params);
      case "sessions.history":
        return this.handleSessionsHistory(params);
      default:
        throw new Error(`Unknown method: ${frame.method}`);
    }
  }

  // ── Method handlers ──────────────────────────────────────────────────

  private async handleChatSend(
    client: GatewayClient,
    params: Record<string, unknown> | undefined,
  ): Promise<{ status: string }> {
    const sessionKey = (params?.sessionKey as string) || "main";
    const message = params?.message as string;
    if (!message) throw new Error("Missing 'message' param");

    await this.opts.sessionManager.resolveOrCreate(sessionKey);
    client.subscribedSession = sessionKey;

    const onEvent: AgentTurnCallback = (event) => {
      const eventFrame: EventFrame = {
        type: "event",
        event: `chat.${event.type}`,
        payload: { sessionKey, ...event },
        seq: this.eventSeq++,
      };
      const json = JSON.stringify(eventFrame);
      console.log(`[gateway] -> Broadcasting event:`, json);
      // Broadcast to all clients subscribed to this session
      for (const c of this.clients.values()) {
        if (c.subscribedSession === sessionKey && c.ws.readyState === WebSocket.OPEN) {
          c.ws.send(json);
        }
      }
    };

    // Fire and forget — the response events stream back via WS events
    console.log("[gateway] Handing off chat to agentRuntime:", { sessionKey, message });
    void this.opts.agentRuntime.runTurn(sessionKey, message, onEvent).catch((err) => {
      onEvent({ type: "error", error: err instanceof Error ? err.message : String(err) });
    });

    return { status: "queued" };
  }

  private async handleSessionsList(): Promise<{ sessions: unknown[] }> {
    return { sessions: this.opts.sessionManager.listSessions() };
  }

  private async handleSessionsCreate(
    params: Record<string, unknown> | undefined,
  ): Promise<unknown> {
    const sessionKey = (params?.sessionKey as string) || "main";
    const title = params?.title as string | undefined;
    return this.opts.sessionManager.createSession(sessionKey, title);
  }

  private async handleSessionsDelete(
    params: Record<string, unknown> | undefined,
  ): Promise<{ deleted: boolean }> {
    const sessionKey = params?.sessionKey as string;
    if (!sessionKey) throw new Error("Missing 'sessionKey' param");
    const deleted = await this.opts.sessionManager.deleteSession(sessionKey);
    return { deleted };
  }

  private async handleSessionsResolve(
    params: Record<string, unknown> | undefined,
  ): Promise<unknown> {
    const sessionKey = (params?.sessionKey as string) || "main";
    return this.opts.sessionManager.resolveOrCreate(sessionKey);
  }

  private async handleSessionsHistory(
    params: Record<string, unknown> | undefined,
  ): Promise<{ messages: unknown[] }> {
    const sessionKey = (params?.sessionKey as string) || "main";
    const messages = await this.opts.sessionManager.readTranscript(sessionKey);
    return { messages };
  }

  // ── Helpers ──────────────────────────────────────────────────────────

  private sendError(ws: WebSocket, id: string, code: string, message: string): void {
    const res: ResponseFrame = {
      type: "res",
      id,
      ok: false,
      error: { code, message },
    };
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify(res));
    }
  }

  private broadcastEvent(event: string, payload: unknown): void {
    const frame: EventFrame = {
      type: "event",
      event,
      payload,
      seq: this.eventSeq++,
    };
    const json = JSON.stringify(frame);
    for (const client of this.clients.values()) {
      if (client.ws.readyState === WebSocket.OPEN) {
        client.ws.send(json);
      }
    }
  }
}
